src/main/java/org/apache/kafka/server/quota/StaticQuotaCallback.java

```diff
@@ -29,6 +42,11 @@
 
     @Override
     public Double quotaLimit(ClientQuotaType quotaType, Map<String, String> metricTags) {
+        // Don't allow producing messages if we're beyond the storage limit.
+        if (ClientQuotaType.PRODUCE.equals(quotaType) && storageUsed.get() > storageQuota) {
+            log.info("Limting producer limit because disk is full. Used: {}. Actual: {}", storageUsed.get(), storageQuota);
+            return 1.0;
```

Comment:
It might be nicer to gradually slow the producer (e.g. linearly between a soft and a hard threshold), which would at least give the user more of an opportunity to realise something was going wrong before their workload just stopped working. I wonder also about adding a metric to this plugin, so that it could get out to prometheus and set of an alert (there should, of course, be altering on the disk usage already, but there's a difference between "You're starting to run out of disk space, but there are no consequences right now" and "Right we're started to slow your producers because you didn't listen the first time" to "Well now you're getting no throughput but it's your own fault".

Reply:
Agreed on both points.